### PR TITLE
Bug 1552511 - Use finalizers to destruct metrics

### PR DESCRIPTION
**Pull request by badboy**
https://bugzilla.mozilla.org/show_bug.cgi?id=1552511

### Pull Request checklist ###

- [ ] **Quality**: This PR builds and tests run cleanly
  - `make test` runs without emitting any warnings
  - `make lint` runs without emitting any errors
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes a changelog entry or an explanation of why it does not need one
  - Any breaking changes to language binding APIs are noted explicitly

### CODE REVIEW

File: glean-core/android/src/main/java/mozilla/telemetry/glean/private/BooleanMetricType.kt [hunk 1]
```diff
@@ -32,6 +33,13 @@ class BooleanMetricType(
                 disabled = disabled.toByte())
     }
 
+    protected fun finalize() {
```
Comment (Dexterp37) [inline glean-core/android/src/main/java/mozilla/telemetry/glean/private/BooleanMetricType.kt]: @badboy, is this a Rust specific behaviour/trait? When does this get called?

Comment (Dexterp37) [reply glean-core/android/src/main/java/mozilla/telemetry/glean/private/BooleanMetricType.kt]: Oh wait, this is Kotlin code. So it must be what Mike mentioned [here](https://kotlinlang.org/docs/reference/java-interop.html#finalize). But that should be for imported Java object: did you check this works?

Comment (badboy) [reply glean-core/android/src/main/java/mozilla/telemetry/glean/private/BooleanMetricType.kt]: Kinda. I added a print statement and run the tests and checked the stdout log to contain the finalize thing, so yeah, it gets called (and didn't crash).

Of course finalizers are not guaranteed to run (or its at least nondeterministic _when_ they run).

**Comment by badboy** [comment]
> When I filed the bug, I considered it pretty low-priority in practice because these objects are all essentially static lifetime singletons (except in unit tests), but I like the idea of being "correct" here.

Yeah, I spent ~5 minutes after clicking on the link in the bug and considered it worth doing earlier today to start into the day :D

**Pull request merged**